core/src/main/java/org/apache/iceberg/formats/FormatModelRegistry.java

```diff
@@ -0,0 +1,237 @@
+ *   <li>{@link EqualityDeleteWriteBuilder} for equality deletes,
+ *   <li>{@link PositionDeleteWriteBuilder} for position deletes.
+ * </ul>
+ *
+ * The appropriate builder is selected based on {@link FileFormat} and object model name.
+ *
+ * <p>{@link FormatModel} objects are registered through {@link #register(FormatModel)} and used for
+ * creating readers and writers. Read builders are returned directly from the factory. Write
+ * builders may be wrapped in specialized content file writer implementations depending on the
+ * requested builder type.
+ */
+public final class FormatModelRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(FormatModelRegistry.class);
+  // The list of classes which are used for registering the reader and writer builders
+  private static final List<String> CLASSES_TO_REGISTER =
+      ImmutableList.of(
+          "org.apache.iceberg.data.GenericFormatModels",
+          "org.apache.iceberg.arrow.vectorized.ArrowReader",
+          "org.apache.iceberg.flink.data.FlinkFormatModels",
+          "org.apache.iceberg.spark.source.SparkFormatModels");
+
+  // Format models indexed by file format and object model class
+  private static final Map<Pair<FileFormat, Class<?>>, FormatModel<?, ?>> MODELS =
+      Maps.newConcurrentMap();
+
+  private FormatModelRegistry() {}
+
+  /**
+   * Registers an {@link FormatModel} in this registry.
+   *
+   * <p>The {@link FormatModel} creates readers and writers for a specific combinations of file
+   * format (Parquet, ORC, Avro) and object model (for example: "generic", "spark", "flink", etc.).
+   * Registering custom factories allows integration of new data processing engines for the
+   * supported file formats with Iceberg's file access mechanisms.
+   *
+   * <p>Each factory must be uniquely identified by its combination of file format and object model
+   * name. This uniqueness constraint prevents ambiguity when selecting factories for read and write
+   * operations.
+   *
+   * @param formatModel the factory implementation to register
+   * @throws IllegalArgumentException if a factory is already registered for the combination of
+   *     {@link FormatModel#format()} and {@link FormatModel#type()}
+   */
+  public static synchronized void register(FormatModel<?, ?> formatModel) {
+    Pair<FileFormat, Class<?>> key = Pair.of(formatModel.format(), formatModel.type());
+
+    FormatModel<?, ?> existing = MODELS.get(key);
+    Preconditions.checkArgument(
+        existing == null || equals(existing, formatModel),
+        "Cannot register %s: %s is registered for format=%s type=%s schemaType=%s",
+        formatModel.getClass(),
+        existing == null ? null : existing.getClass(),
+        key.first(),
+        key.second(),
+        existing == null ? null : existing.schemaType());
+
+    MODELS.put(key, formatModel);
+  }
+
+  private static boolean equals(FormatModel<?, ?> model1, FormatModel<?, ?> model2) {
```

Comment:
Style: Avoid sandwiching private methods between public methods. Usually, put private methods at the end of the file.

Reply:
Reordered the methods